receiver/hostmetricsreceiver/internal/scraper/cpuscraper/internal/metadata/generated_metrics.go

```diff
@@ -24,75 +24,118 @@ import (
+//
+// Typical use case:
+// m := CpuMetric.Init(startTime)
+// m.EnsureDataPointsCapacity(1)
+// m.Record(now, value, "cpu0", "idle")
+// m.AppendToMetricSlice(metrics)
+type CpuMetric struct {
+	mb             CpuMetricBuilder
+	metric         pdata.Metric
+	startTimestamp pdata.Timestamp
 }
 
-// Intentionally not exposing this so that it is opaque and can change freely.
-type metricImpl struct {
-	name     string
-	initFunc func(pdata.Metric)
+func (m CpuMetric) EnsureDataPointsCapacity(cap int) {
+	if !m.mb.Enabled() {
+		return
+	}
+	m.metric.Sum().DataPoints().EnsureCapacity(cap)
+	// For other types:
+	// mt.metric.Gauge().DataPoints().EnsureCapacity(cap)
+	// mt.metric.Histogram().DataPoints().EnsureCapacity(cap)
+	// mt.metric.Summary().DataPoints().EnsureCapacity(cap)
 }
 
-// Name returns the metric name.
-func (m *metricImpl) Name() string {
-	return m.name
+// Record adds a data point to CpuMetric.
+// If provided attribute is of AttributeValueTypeEmpty type, it will be skipped.
+func (m CpuMetric) Record(ts pdata.Timestamp, val float64, cpuAttributeValue, stateAttributeValue pdata.AttributeValue) {
+	if !m.mb.Enabled() {
+		return
+	}
+
+	dp := m.metric.Sum().DataPoints().AppendEmpty()
+	dp.SetStartTimestamp(m.startTimestamp)
+	dp.SetTimestamp(ts)
+	dp.SetDoubleVal(val)
+	if cpuAttributeValue.Type() != pdata.AttributeValueTypeEmpty {
+		dp.Attributes().Insert(L.Cpu, cpuAttributeValue)
+	}
+	if stateAttributeValue.Type() != pdata.AttributeValueTypeEmpty {
+		dp.Attributes().Insert(L.State, stateAttributeValue)
+	}
 }
 
-// New creates a metric object preinitialized.
-func (m *metricImpl) New() pdata.Metric {
-	metric := pdata.NewMetric()
-	m.Init(metric)
-	return metric
+// AppendToMetricSlice appends CpuMetric to a pdata.MetricsSlice.
+func (m CpuMetric) AppendToMetricSlice(metrics pdata.MetricSlice) {
+	if !m.mb.Enabled() {
+		return
+	}
+	m.metric.CopyTo(metrics.AppendEmpty())
 }
 
-// Init initializes the provided metric object.
-func (m *metricImpl) Init(metric pdata.Metric) {
-	m.initFunc(metric)
+type CpuMetricBuilder struct {
+	name        string
+	enabled     bool
+	description string
+	unit        string
+	dataType    pdata.MetricDataType
+	isMonotonic bool
+	temporality pdata.MetricAggregationTemporality
+	config      MetricConfig
 }
 
-type metricStruct struct {
-	SystemCPUTime MetricIntf
+// Name returns the metric name.
+func (mb CpuMetricBuilder) Name() string {
+	return mb.name
 }
 
-// Names returns a list of all the metric name strings.
-func (m *metricStruct) Names() []string {
-	return []string{
-		"system.cpu.time",
+// Enabled identifies whether the metrics should be collected or not.
+func (mb CpuMetricBuilder) Enabled() bool {
+	if mb.config.Enabled != nil {
+		return *mb.config.Enabled
 	}
+	return mb.enabled
 }
 
-var metricsByName = map[string]MetricIntf{
-	"system.cpu.time": Metrics.SystemCPUTime,
+// Init generates CpuMetric.
+func (mb CpuMetricBuilder) Init(startTimestamp pdata.Timestamp) CpuMetric {
```

Comment:
I think we can keep track of "startTime" between calls here so no need for the scraper to do that? Otherwise we need to pass it with every point, because the source may track that per timeseries/point? 

We need to think if this can change while a "scraper" is alive, I think it can change when let's say we scrape "redis" and it restarts. Do we have any use-case/example where startTime changes at point level?

Reply:
For example `kubelet` receiver takes the whole set of stats from kubelet API including metric's startTime, that should stay the same but we cannot assume that.

I'd like to support changing the startTime as an "optional" way, but cannot think of any good way to do this. Two options comes up to my mind, but I don't feel good about any of them:
- Providing a `RecordWithStartTime` method in addition to `Record`. 
- Providing a `SetStartTimestamp` method that would update the builder's state.

May be we should keep passing it around instead. What do you think?

Reply:
The current approach in this PR where we report startTime on every scrape works for kubelet and should work for others I believe.

Reply:
I think we should not block on this. I need to think about more cases, and probably we need to use this API in more places before we know for sure. What we can do is to accept startTime for the moment in Record, and after we change 2-3 scrapers (one of the host, maybe kublet and maybe redis/mysql) we revisit this.

We cannot get it perfect right now because there are lots of unknown things, let's iterate, so for the moment keep it with every record.